add document for v5.1: Get Started -> Features (#230)
tikv/website: content/docs/5.1/concepts/explore-tikv-features/cas.md

---
title: CAS on RawKV
description: Compare And Swap
menu:
    "5.1":
        parent: Features
        weight: 4
---

This page walks you through a simple demonstration of how to use TiKV's `compare-and-swap`.

In RawKV, `compare-and-swap (CAS)` is an atomic instruction used in multi-threading to achieve synchronization. It compares the `current_value` of a `key` with a given value and, only if they are the same, modifies the contents of that `key` to a new given value. This is done as a single atomic operation. The atomicity guarantees that the new value is calculated based on up-to-date information; if the value had been updated by another thread in the meantime, the write would fail.

> ## Prerequisites

Reviewer: I don't see anywhere below we start a TiKV cluster by TiUP. And I think `cargo run` requires an existing cluster.

Author: thx, i forget that.

Reviewer 2: This is a section to describe how to set up the tikv cluster in the before, you only need to reference that document, no need to write the same content again.

1. Install TiUP by executing the following command:

```bash
curl --proto '=https' --tlsv1.2 -sSf https://tiup-mirrors.pingcap.com/install.sh | sh
```

2. If TiUP is already installed, update the TiUP playground component to the latest version:

```bash
tiup update --self && tiup update playground
```

3. Install Rust

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

## Step 1: Create a rust project

Use `cargo` to create a rust project:

```bash
cargo new hello-tikv
cd hello-tikv
```

## Step 2: Add dependency

Append the following content to file `Cargo.toml`.

```toml
tikv-client = "0.1.0"
futures = "0.3"
```

## Step 3: Write test code

Save the following code to file `src/main.rs`.

```rust
use tikv_client::RawClient;
use futures::executor::block_on;

fn main() {

>     block_on(raw_cas());

Reviewer 3: Could we use `#[tokio::main]` instead?

Author: sure

}

async fn raw_cas() {
    let client = RawClient::new(vec!["127.0.0.1:2379"])
        .await.unwrap()

>         // Users must set `with_atomic_for_cas` in `put`, `delete`,

Reviewer: What do you want to express by "in"? It seems a client level configuration.

        // `batch_delete` and `batch_put` to ensure linearizability
        // of `compare_and_swap`.
        .with_atomic_for_cas();

    let key = "key".to_owned();
    let value = "value".to_owned();
    let new_value = "newValue".to_owned();

    // call put(key, value)
    client.put(key.clone(), value.clone()).await.unwrap();
    println!("put key={}, value={}", key, value);

    // call get(key) returns value
    let res = client.get(key.clone()).await.unwrap();
    println!("get key={} returnedValue {:?}", key, res);

    // call compare_and_swap(key, value, new_value) returns (value, true)
    // If the value retrived is equal to `current_value`, `new_value` is written.
    // A tuple is returned if successful: the previous value and whether the value is swapped
    let res = client
        .compare_and_swap(

>             key.clone(),

Reviewer 3: could we use reference instead of clone?

Author: use `to_owned` instead

Reviewer 3: `Key` implement this trait https://docs.rs/tikv-client/0.1.0/src/tikv_client/kv/key.rs.html#160-162, so and `&Vec<u8>` could be use as argument in `compare_and_swap`

            Some(value.to_owned()).map(|v| v.into()),
            new_value.clone(),
        )
        .await.unwrap();

>     println!("compare_and_swap key={} oldValue={} newValue={} returnedValue={:?}", key, value, new_value, res);

Reviewer 3: use assertion would look more like Rust example

Reviewer 3: here is an example from [sled](https://github.com/spacejam/sled)

Author: add `assert_eq`

## Prerequisites

1. Install TiUP by executing the following command:

```bash
curl --proto '=https' --tlsv1.2 -sSf https://tiup-mirrors.pingcap.com/install.sh | sh
```

2. If TiUP is already installed, update the TiUP playground component to the latest version:

```bash
tiup update --self && tiup update playground
```

3. Install Rust

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

## Step 1: Start TiKV Cluster

For the purpose of this tutorial, you need only one TiKV node, so use the `tiup playground` command.

Show TiUP version:

```bash
tiup -v
```

version >= 1.5.0:

```bash
tiup playground --mode tikv-slim
```

version < 1.5.0:

```bash
tiup playground
```

## Step 2: Create a rust project

Use `cargo` to create a rust project:

```bash
cargo new hello-tikv
cd hello-tikv
```

## Step 3: Add dependency

Append the following content to file `Cargo.toml`.

```toml
tikv-client = "0.1.0"
tokio = { version = "1.5.0", features = ["full"] }
```

> ## Step 4: Write test code

Reviewer 2: maybe it's better to use Java as the first language for examples since it's most widely used in application developments.

Author: As discussed, RawKV's examples will use `java`. 
So I will change the example of `CAS` from `rust` to `java`.

Author: Can we continue to review the PR? I will update the example as soon as the CAS API is ready in java client.

Reviewer 2: OK.

Save the following code to file `src/main.rs`.

```rust
use tikv_client::RawClient;

#[tokio::main]
async fn main() {
    let client = RawClient::new(vec!["127.0.0.1:2379"])
        .await.unwrap()

>         // Users must set `with_atomic_for_cas` when using `put`, `delete`,
>         // `batch_delete` and `batch_put` to ensure linearizability
>         // of `compare_and_swap`.

Reviewer: ```suggestion
        // Users must set `with_atomic_for_cas` to ensure linearizability of 
        // `compare_and_swap` when used together with `put`, `delete`,
        // `batch_put`, or `batch_delete`
```

    println!("get key={} returnedValue {:?}", key, res);

    // call compare_and_swap(key, value, new_value) returns (value, true)
    // If the value retrived is equal to `current_value`, `new_value` is written.
    // A tuple is returned if successful: the previous value and whether the value is swapped
    let res = client
        .compare_and_swap(
            key.to_owned(),
            Some(value.to_owned()).map(|v| v.into()),
            new_value.to_owned(),
        )
        .await.unwrap();
    println!("compare_and_swap key={} oldValue={} newValue={} returnedValue={:?}", key, value, new_value, res);

    // call get(key) returns new_value
    let res = client.get(key.to_owned()).await.unwrap();
    println!("get key={} returnedValue={:?}", key, res);
}
```

## Step 5: Run the code

```bash
cargo run

put key=key, value=value
get key=key returnedValue Some([118, 97, 108, 117, 101])
compare_and_swap key=key oldValue=value newValue=newValue returnedValue=(Some([118, 97, 108, 117, 101]), true)
get key=key returnedValue=Some([110, 101, 119, 86, 97, 108, 117, 101])
```

As we can see, after calling `compare_and_swap` the `value` is replaced by `newValue`.

> {{< warning >}}

Reviewer 3: Seems this warning is already in example code comments? Could we remove this and add a **NOTE** header in the comments?

Author: It's a bit repetitive. I remove the comments in code.

> Users must set `with_atomic_for_cas` in `put`, `delete`, `batch_delete` and `batch_put` to ensure linearizability of `compare_and_swap`.

Reviewer: ```suggestion
Users must set `with_atomic_for_cas` to ensure linearizability of `compare_and_swap` when used together with `put`, `delete`, `batch_put`, or `batch_delete`
```

---
title: CAS on RawKV
description: Compare And Swap
menu:
    "5.1":
        parent: Features
        weight: 4
---

This page walks you through a simple demonstration of how to use TiKV's `compare-and-swap`.

In RawKV, `compare-and-swap (CAS)` is an atomic instruction used in multi-threading to achieve synchronization. It compares the `current_value` of a `key` with a given value and, only if they are the same, modifies the contents of that `key` to a new given value. This is done as a single atomic operation. The atomicity guarantees that the new value is calculated based on up-to-date information; if the value had been updated by another thread in the meantime, the write would fail.

> ## Prerequisites

Reviewer 2: no need to install tiup and start the tikv cluster again, since it already expressed in https://tikv.org/docs/5.1/concepts/tikv-in-5-minutes/

Let's leave the content to how to use CAS in Java and Rust
